core/src/Kokkos_Parallel_Reduce.hpp

```diff
@@ -1359,29 +1359,51 @@ struct ParallelReduceAdaptor {
                                   const PolicyType& policy,
                                   const FunctorType& functor,
                                   ReturnType& return_value) {
-    uint64_t kpID = 0;
+    using PassedReducerType = typename return_value_adapter::reducer_type;
+    uint64_t kpID           = 0;
 
     PolicyType inner_policy = policy;
-    Kokkos::Tools::Impl::begin_parallel_reduce<
-        typename return_value_adapter::reducer_type>(inner_policy, functor,
-                                                     label, kpID);
+    Kokkos::Tools::Impl::begin_parallel_reduce<PassedReducerType>(
+        inner_policy, functor, label, kpID);
 
     Kokkos::Impl::shared_allocation_tracking_disable();
-    Impl::ParallelReduce<FunctorType, PolicyType,
-                         typename return_value_adapter::reducer_type>
-        closure(functor, inner_policy,
-                return_value_adapter::return_value(return_value, functor));
-    Kokkos::Impl::shared_allocation_tracking_enable();
-    closure.execute();
-
-    Kokkos::Tools::Impl::end_parallel_reduce<
-        typename return_value_adapter::reducer_type>(inner_policy, functor,
-                                                     label, kpID);
+
+    using ReducerType =
+        std::conditional_t<std::is_same_v<InvalidType, PassedReducerType>,
+                           FunctorType, PassedReducerType>;
+    using Analysis =
+        FunctorAnalysis<FunctorPatternInterface::REDUCE, PolicyType,
+                        ReducerType, typename return_value_adapter::value_type>;
+    using ParallelReduceType =
+        Impl::ParallelReduce<FunctorType, PolicyType,
+                             typename Analysis::Reducer,
+                             typename Impl::FunctorPolicyExecutionSpace<
+                                 FunctorType, PolicyType>::execution_space>;
+
+    if constexpr (std::is_same_v<InvalidType, PassedReducerType>) {
+      typename Analysis::Reducer final_reducer(functor);
+      ParallelReduceType closure(
+          functor, inner_policy, final_reducer,
+          return_value_adapter::return_value(return_value, functor));
+      Kokkos::Impl::shared_allocation_tracking_enable();
+      closure.execute();
+    } else {
+      auto reducer = return_value_adapter::return_value(return_value, functor);
+      typename Analysis::Reducer final_reducer(reducer);
```

Comment:
Are you potentially pass two functor copies on below? One for the reducer one for the functor? Are both of those reaching the device?

Reply:
Yes, there are two functor copies if there is no reducer. 
If that is an issue, we would need to combine the functor and the `FunctorAnalysis` object into a different class that we would pass to the `ParallelReduce` implementations.

Reply:
Yeah that is definitely an issue. It effectively halfs the size available to hit the local_launch_path in CUDA/HIP and halfs the size available for constant cache usage.

Reply:
I'll look into passing a wrapper for both the functor and the reducer to avoid having to copies (which will of course make this pull request even bigger).
If we just want to move the deduction of the value size from the functor signature to the return type first (to avoid doing to make changes in one pull request), we could just do https://github.com/kokkos/kokkos/pull/5169 first (even though changes to `ParallelReduce`'s template parameters will be reverted in this pull request again).

Reply:
I added the wrapper in [`1c97d20` (#5332)](https://github.com/kokkos/kokkos/pull/5332/commits/1c97d20a84b9ae544928057dc2f7bd16d87f5410) and subsequent commits.